### PR TITLE
feat(controller): Add a shared index informer for ConfigMaps

### PR DESCRIPTION
**Pull request by terrytangyuan**
This PR adds the informer for ConfigMaps based on our discussion. Let me know if this aligns with what you had in mind.

cc @alexec @sarabala1979 

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -306,6 +313,8 @@ func (wfc *WorkflowController) startLeading(ctx context.Context, logCtx *log.Ent
 	for i := 0; i < podWorkers; i++ {
 		go wait.Until(wfc.podWorker, time.Second, ctx.Done())
 	}
+
+	go wait.UntilWithContext(ctx, wfc.syncAllConfigMaps, 10*time.Second)
```
Comment (alexec) [inline workflow/controller/controller.go:317]: every 10s - I would expect this to cause too many allocs and result in OOMKilled - why do we need this?

File: workflow/controller/controller.go [hunk 2]
```diff
@@ -1062,6 +1088,15 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) cache.SharedI
 	return informer
 }
 
+func (wfc *WorkflowController) newConfigMapInformer() cache.SharedIndexInformer {
+	labelSelector, _ := labels.Parse(indexes.ConfigMapTypeLabel)
+	return v1.NewFilteredConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), configMapResyncPeriod, cache.Indexers{
+		indexes.ConfigMapLabelsIndex: indexes.ConfigMapIndexFunc,
+	}, func(opts *metav1.ListOptions) {
+		opts.LabelSelector = labelSelector.String()
```
Comment (alexec) [inline workflow/controller/controller.go:1096]: minor - I find that the Goclient code for labels is really hard to use and opaque, I'd like inline the selector as a string

File: workflow/controller/indexes/configmap_index.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package indexes
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+const ConfigMapTypeLabel = "workflows.argoproj.io/configmap-type"
+
+func ConfigMapIndexFunc(obj interface{}) ([]string, error) {
```
Comment (alexec) [inline workflow/controller/indexes/configmap_index.go:9]: 👍

File: workflow/controller/indexes/configmap_index_test.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestConfigMapIndexFunc(t *testing.T) {
```
Comment (alexec) [inline workflow/controller/indexes/configmap_index_test.go:12]: 👍

File: workflow/controller/controller.go [hunk 5]
```diff
@@ -117,6 +120,7 @@ const (
 	workflowTemplateResyncPeriod        = 20 * time.Minute
 	podResyncPeriod                     = 30 * time.Minute
 	clusterWorkflowTemplateResyncPeriod = 20 * time.Minute
+	configMapResyncPeriod               = 20 * time.Minute
```
Comment (alexec) [inline workflow/controller/controller.go:123]: inline this if used just once

File: workflow/controller/cache/configmap_cache.go [hunk 6]
```diff
@@ -117,6 +119,7 @@ func (c *configMapCache) Save(ctx context.Context, key string, nodeId string, va
 	}
 
 	creationTime := time.Now()
+	cache.SetLabels(map[string]string{indexes.ConfigMapTypeLabel: "cache"})
```
Comment (alexec) [inline workflow/controller/cache/configmap_cache.go:122]: I don't think we want this yet?

**Review by alexec — CHANGES_REQUESTED** [review]

File: workflow/controller/controller.go (hunk 1 above)
Comment (terrytangyuan) [reply workflow/controller/controller.go:317]: This is to prepare future work, e.g. processing of individual configmaps and GC when condition is met. Any suggestions? I can change this to less frequent, e.g. 1 minute and make it configurable via env var.

Comment (terrytangyuan) [reply workflow/controller/controller.go]: I am going to remove this part in this PR and add later when needed.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
@sarabala1979 this is some pre-work that @terrytangyuan is doing for one of the top voted issues.

**Pull request merged**